include/nuttx/can/sja1000.h

```diff
@@ -0,0 +1,163 @@
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+#ifndef __DRIVERS_INCLUDE_CAN_SJA1000_H
+#define __DRIVERS_INCLUDE_CAN_SJA1000_H
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <nuttx/config.h>
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Public Types
+ ****************************************************************************/
+
+/* CAN hardware-dependent bit-timing constant
+ *
+ * Used for calculating and checking bit-timing parameters
+ */
+
+struct sja1000_can_bittiming_const
+{
+  char name[16];      /* Name of the CAN controller hardware */
+  uint32_t tseg1_min; /* Time segment 1 = prop_seg + phase_seg1 */
+  uint32_t tseg1_max;
+  uint32_t tseg2_min; /* Time segment 2 = phase_seg2 */
+  uint32_t tseg2_max;
+  uint32_t sjw_min; /* Synchronization jump width */
+  uint32_t sjw_max; /* Synchronisation jump width */
+  uint32_t brp_min; /* Bit-rate prescaler */
+  uint32_t brp_max;
+  uint32_t brp_inc;
+};
+
+/* Type of the SJA1000 interrupt handling callback */
+
+struct sja1000_config_s; /* Forward reference */
+typedef CODE void (*sja1000_handler_t)(FAR struct sja1000_config_s *config,
+                                       FAR void *arg);
+
+/* A reference to a structure of this type must be passed to the SJA1000
+ * driver when the driver is instantiated. This structure provides
+ * information about the configuration of the SJA1000 and provides some
+ * board-specific hooks.
+ *
+ * Memory for this structure is provided by the caller.  It is not copied by
+ * the driver and is presumed to persist while the driver is active. The
+ * memory must be writeable because, under certain circumstances, the driver
+ * may modify the frequency.
+ */
+
+struct sja1000_config_s
+{
+  /* Device configuration */
+
+  const struct sja1000_can_bittiming_const *bittiming_const;
+  uint8_t port;       /* SJA1000 device port number */
+  uint8_t periph;     /* Peripheral ID (optional) */
+  uint8_t irq;        /* IRQ associated with this SJA1000 */
+  uint8_t cpu;        /* CPU ID */
+  int8_t cpuint;      /* CPU interrupt assigned to this SJA1000 */
+  uint32_t bitrate;   /* Configured bit rate */
+  uint32_t samplep;   /* Configured sample point */
+  uint32_t sjw;       /* Synchronization jump width */
+  uint32_t base;      /* SJA1000 register base address */
```

Comment:
The `base` is required in each putreg and getreg and as I suggest to move its addition into actual functions, I suggest to copy value from `struct sja1000_config_s` into same field name introduced in `struct sja1000_dev_s` during initialization. It prevents double referencing
```
priv->config->base
```